src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs

```diff
@@ -0,0 +1,322 @@
+            : this()
+        {
+            if (functions == null)
+            {
+                throw new ArgumentNullException(nameof(functions));
+            }
+
+            foreach (var func in functions)
+            {
+                Add(func);
+            }
+        }
+
+        internal TexlFunctionSet(IEnumerable<TexlFunctionSet> functionSets)
+            : this()
+        {
+            if (functionSets == null)
+            {
+                throw new ArgumentNullException(nameof(functionSets));
+            }
+
+            foreach (var functionSet in functionSets)
+            {
+                if (functionSet._count > 0)
+                {
+                    Add(functionSet);
+                }
+            }
+        }
+
+        // Slow API, only use for backward compatibility
+        [Obsolete("Slow API. Prefer using With* functions for identifying functions you need.")]
+        internal IEnumerable<TexlFunction> Functions
+        {
+            get
+            {
+                foreach (var kvp in _functions)
+                {
+                    foreach (var func in kvp.Value)
+                    {
+                        yield return func;
+                    }
+                }
+            }
+        }
+
+        internal TexlFunction Append(TexlFunction function) => Add(function);
+
+        internal TexlFunction Add(TexlFunction function)
+        {
+            if (function == null)
+            {
+                throw new ArgumentNullException(nameof(function));
+            }
+
+            var fList = WithNameInternal(function.Name);
+
+            if (fList.Any())
+            {
+                if (fList.Contains(function))
+                {
+                    throw new ArgumentException($"Function {function.Name} is already part of core or extra functions");
+                }
+
+                fList.Add(function);
+            }
+            else
+            {
+                _functions.Add(function.Name, new List<TexlFunction>() { function });
+            }
+
+            var fInvariantList = WithInvariantNameInternal(function.LocaleInvariantName);
+
+            if (fInvariantList.Any())
+            {
+                if (fInvariantList.Contains(function))
+                {
+                    throw new ArgumentException($"Function {function.Name} is already part of core or extra functions (invariant)");
+                }
+
+                fInvariantList.Add(function);
+            }
+            else
+            {
+                _functionsInvariant.Add(function.LocaleInvariantName, new List<TexlFunction>() { function });
+            }
+
+            var fnsList = WithNamespaceInternal(function.Namespace);
+
+            if (fnsList.Any())
+            {
+                if (fnsList.Contains(function))
+                {
+                    throw new ArgumentException($"Function {function.Name} is already part of core or extra functions (namespace)");
```

Comment:
>throw new ArgumentException($"Function {function.Name} is already part of core or extra functions (namespace)"); [](http://example.com/codeflow?start=20&length=112)

Should we worry about side effect, if this throws we don't remove from _functionsInvariantand & _functions? similarly for above throw?

Reply:
Thinking of it ... I fixed the code ;)